### ISSUE TITLE
Provide an Error Analysis of INT8 Quantization

### ISSUE THREAD

**Issue by JosephChenHub** [ultralytics/yolov5#2509]
https://github.com/NVIDIA/TensorRT/issues/1114

**Comment by glenn-jocher** [comment]
@JosephChenHub thanks for the excellent study. Posting here in case anyone is interested:

> ## Description
> When doing INT8 PTQ (post-training quantization), we notice that the dropout of the accuracy on image classification or semantic segmentation is rather small. However, for the detectors like yolov5, there exists a huge gap of mAP between FP16 model and INT8 model. The comparison results are listed as the following, e.g., for [YOLOv5s](https://github.com/ultralytics/yolov5/releases), the mAP on COCO validation set decreases from 0.362 to 0.054 (or 0.316, 0.054 is the result of original onnx graph and 0.316 the result of custom plugin of `SiLU`).
> 
> ![image](https://user-images.githubusercontent.com/22444830/110743161-8c7fcb00-8272-11eb-958b-77c8dca90e8f.png)
> 
> The huge gap between 0.362 and 0.054 makes the model unreliable, and we observed that the yolov5s model utilizes an activation function `SiLU` (SiLU(x) = x * sigmoid(x)) which may reduce the accuracy due to its nonlinearity.
> ![image](https://user-images.githubusercontent.com/22444830/110743971-e634c500-8273-11eb-9494-8c0ebc3dfb7d.png)
> 
> Hence, we manually write a plugin `SiLU` to replace the origin operator `sigmoid` and `mul` in the onnx graph, which looks like
> ![silu](https://user-images.githubusercontent.com/22444830/110744129-285e0680-8274-11eb-8f24-0c4941ff0af7.png)
> ![silu2](https://user-images.githubusercontent.com/22444830/110744141-2bf18d80-8274-11eb-89b1-a5bee5c10e47.png)
> 
> As a result, we can see from the above table that our detector keeps stable and there still exists a little gap between the FP16 models and the INT8 model (for yolov5s, the mAP/FPS improves from 0.054/464 to 0.316/327 due to our custom plugin). Furthermore, we do partial quantization (use top 5 sensitive conv. layers) and the mAP is 0.332, which is still lower than 0.362.
> So questions come that
> 
> 1. In the origin implementation, we see the log `Layer(PointWiseV2): PWN(Sigmoid_213, Mul_214), Tactic: 24, 345[Int8(128,40,40)] -> 347[Int8(128,40,40)]` , and how the TensorRT implements the INT8 operation on this layer (sigmoid + mul) ?
> 2. In our own custom plugin, we implement the `SiLU` as
> 
> ```
> template <typename T>
> __device__ T _silu_op(const T input);
> 
> template <>
> __device__  __forceinline__ float _silu_op(const float input) { 
>     return input / (static_cast<float>(1) + __expf(-input));
> }
> template <>
> __device__  __forceinline__ __half _silu_op(const __half input) {
>     typedef __half T;
>     return input / (static_cast<T>(1) + hexp(-input));
> }
> 
> template <typename T>
> __global__ void _kernel(const T* __restrict__ inputs, T* __restrict__ outputs, const size_t n ) { 
>     const size_t tid = threadIdx.x + blockIdx.x * blockDim.x;
>     if (tid < n) {
>         outputs[tid] = _silu_op<T>(inputs[tid]);
>     }   
> }
> ```
> 
> does there exist a faster way to achieve this op? or TensorRT will optimize this op like `ReLU`, `Sigmoid` in future ?
> 3. We notice that the sensitive layers usually are the first or last several conv. layers in the CNN, and the nonlinear modules are also sensitive. Is the error of PTQ mainly caused by the nonlinearity when doing linear quantization? How to prevent that ? As far as we know, the QAT (quantization aware training) of TensorRT is still not applicable. The dropout of accuracy makes the QAT more desirable. When will TensorRT support QAT well ?
> 
> ## Environment
> **TensorRT Version**: 7.2.1
> **NVIDIA GPU**: 2080 Ti
> **NVIDIA Driver Version**: 10.2
> **CUDA Version**: 10.2
> **CUDNN Version**: 8.0
> **Operating System**: Ubuntu 18.04
> **Python Version (if applicable)**: 3.6
> **Tensorflow Version (if applicable)**:
> **PyTorch Version (if applicable)**: 1.8.0
> **Baremetal or Container (if so, version)**:
> 
> ## References
> yolov5: https://github.com/ultralytics/yolov5
